### DISCUSSION TITLE
5GMS AF: Traffic identification for Network Assistance and Dynamic Policies

### DISCUSSION THREAD

**Discussion by rjb1000** [5G-MAG/rt-5gms-application-function#79] [General]
# Description

When invoking the *Network Assistance API* or the *Dynamic Policies API* at reference point M5, the Media Session Handler needs to specify exactly which application service data flows (i.e. HTTP sessions) the Network Assistance session or Policy Template instance applies to. To this end, when creating a Network Assistance session (TS 26.512 clause 11.6.3.1) or when instantiating a Policy Template (TS 26.512 clause 11.5.3.1), an array of `ServiceDataFlowDescription` objects is supplied by the Media Session Handler.

This data type is specified in clause 6.4.3.2 of TS 26.512 and the Media Session Handler has two choices for specifying each relevant application service data flow, as detailed below. (This is currently specified ambiguously in TS 26.512, but I think the intention was that _either_ an IP Packet Filter Set _or_ a domain name are specified in each Service Data Flow Description. I have raised 5G-MAG/Standards#74 to clarify this ambiguity in TS 26.512.)

1. An **IP Packet Filter Set** signalled in `ServiceDataFlowDescription.flowDescription`. This can identify the server 3-tuple, or a full 5-tuple or a ToS value, amongst other traffic identification parameters.
   - In this case, the _Npcf_PolicyAuthorization_ service (TS 29.514) is invoked on the PCF and each IP Packet Filter Set is translated into appropriate properties of `AppSessionContext.ascReqData`.
2. A **domain name** string signalled in `ServiceDataFlowDescription.domainName`.
   - In this case, the _PfdManagement_ service (TS 29.122) is instead invoked on the NEF (PFDF) to create/modify a Packet Flow Description with the specified domain name.
      - PFDs are managed under the umbrella of an Application Identifier which is one of the properties of a 5GMS Provisioning Session.
      - PFDs are stored persistently in the UDR.
   - As a consequence, the NEF notifies the SMF of this modification using the _Nnef_PFDManagement_Notify_ service operation (TS 29.551) and the SMF in turn adjusts the dynamic PCC Rules in the UPF accordingly.
      - Support for this service in the SMF is a requirement for PFDs.
   - (The _PfdManagement_ service is actually a legacy service originally intended for northbound exposure of PFDs from the Evolved Packet Core for Machine-Type Communications. In that legacy case, the service consumer was SCS/AS and the service was provided by the Service Capabilty Exposure Function (SCEF). However, clause 4.4.6 of TS 29.522 effectively co-opts this API into the 5G System, redefining the service consumer as the AF and the service provider as the PFD Function of the NEF, which is referred to as "NEF (PFDF)".)
   - This service obviously requires the presence of a NEF in the 5G Core, which is missing from Open5GS.
   - Hence, we cannot currently implement the domain name variant of the 5GMS Service Data Flow Description and the 5GMS AF should return an error if one is included in a `ServiceDataFlowDescription` received from a Media Session Handler.

# Summary of relevant service operations

## Application detection using PFD management in the NEF (PFDF)

OpenAPI YAML stage 3 specifications:

- [TS29551_Nnef_PFDmanagement.yaml](https://forge.3gpp.org/swagger/editor-versions/v3.6.23/?url=https://forge.3gpp.org/rep/all/5G_APIs/raw/REL-17/TS29551_Nnef_PFDmanagement.yaml)
- [TS29122_PfdManagement.yaml](https://forge.3gpp.org/swagger/editor-versions/v3.6.23/?url=https://forge.3gpp.org/rep/all/5G_APIs/raw/REL-17/TS29122_PfdManagement.yaml)

<html><body>
<table cellspacing="0" cellpadding="4" border="1" align="center">
  <tbody>
    <tr>
      <th valign="top" bgcolor="#c0c0c0">Stage 2 definitions</th>
      <th rowspan="1" colspan="2" valign="top" bgcolor="#c0c0c0">Stage 3
        specifications</th>
    </tr>
    <tr>
      <th valign="top" bgcolor="#c0c0c0">TS 23.502 (5G System)</th>
      <th valign="top" bgcolor="#c0c0c0">TS 29.551 (5G System)</th>
      <th valign="top" bgcolor="#c0c0c0">TS 29.122 (EPS)</th>
    </tr>
    <tr>
      <td valign="top"><i>Nnef_PFDManagement_Fetch</i></td>
      <td valign="top"><i>Nnef_PFDManagement_Fetch</i></td>
      <td valign="top" bgcolor="#808080"><i>3gpp-pfd-management_FetchIndPFDManagementTransaction</i><br>
        (Redundant?)</td>
    </tr>
    <tr>
      <td valign="top"><i>Nnef_PFDManagement_Subscribe</i></td>
      <td valign="top"><i>Nnef_PFDManagement_Subscribe</i></td>
      <td valign="top" bgcolor="#808080">(Not specified)</td>
    </tr>
    <tr>
      <td valign="top"><i>Nnef_PFDManagement_Notify</i></td>
      <td valign="top"><i>Nnef_PFDManagement_Notify</i></td>
      <td valign="top" bgcolor="#808080">(Not specified)</td>
    </tr>
    <tr>
      <td valign="top"><i>Nnef_PFDManagement_Unsubscribe</i></td>
      <td valign="top"><i>Nnef_PFDManagement_Unsubscribe</i></td>
      <td valign="top" bgcolor="#808080">(Not specified)</td>
    </tr>
    <tr>
      <td valign="top"><i>Nnef_PFDManagement_Create</i></td>
      <td valign="top" bgcolor="#808080">(Not specified)</td>
      <td valign="top"><i>3gpp-pfd-management_UpdateIndPFDManagementTransaction</i></td>
    </tr>
    <tr>
      <td valign="top"><i>Nnef_PFDManagement_Update</i></td>
      <td valign="top" bgcolor="#808080">(Not specified)<br>
      </td>
      <td valign="top"><i>3gpp-pfd-management_ModifyIndPFDManagementTransaction</i></td>
    </tr>
    <tr>
      <td valign="top"><i>Nnef_PFDManagement_Delete</i></td>
      <td valign="top" bgcolor="#808080">(Not specified)</td>
      <td valign="top"><i>3gpp-pfd-management_DeleteIndPFDManagementTransaction</i></td>
    </tr>
  </tbody>
</table>
</body></html>

**Comment by tlohmar** [comment]
FYI, not only the NEF is impacted when using PFDs. Also other Network Functions needs to support the according procedure at their APIs, specifically the SMF and UDR.
